libuavcan/include/uavcan/util/avl_tree.hpp

```diff
@@ -0,0 +1,387 @@
+/*
+ * AVL Tree implementation.
+ * Copyright (C) 2019 Theodoros Ntakouris <zarkopafilis@gmail.com>
+ */
+
+#ifndef UAVCAN_UTIL_AVL_TREE_HPP_INCLUDED
+#define UAVCAN_UTIL_AVL_TREE_HPP_INCLUDED
+
+#include <cstdlib>
+#include <cassert>
+#include <cmath>
+#include <algorithm>
+#include <functional>
+#include <uavcan/build_config.hpp>
+#include <uavcan/util/templates.hpp>
+#include <uavcan/dynamic_memory.hpp>
+#include <uavcan/debug.hpp>
+
+namespace uavcan 
+{
+template <typename T>
+class UAVCAN_EXPORT AvlTree : Noncopyable 
+{
+protected:
+    struct Node {
+        T* data = UAVCAN_NULLPTR;
+        int16_t h = 1; // initially added as leaf
+        Node* left = UAVCAN_NULLPTR;
+        Node* right = UAVCAN_NULLPTR;
+        Node* equal_keys = UAVCAN_NULLPTR;
+    };
+
+    Node* root_;
+
+private:
+    size_t len_ ;
+
+    static int16_t heightOf(const Node* n) {
+        if (n == UAVCAN_NULLPTR) {
+            return 0;
+        }
+
+        return n->h;
+    }
+
+    void postOrderTraverseRecursively(Node* n, std::function<void(T*&)> forEach) {
```

Comment:
let's not put code in an API that is only used for test. What I reccommend is that you write the postOrderTraverseRecursively and walkPostOrder functions into the unit tests. There are a few techniques you c an use for this.
1. **test friend** - This is my preferred method: you declare a friend of your class that is `friend class AvlTreeTestFriend`. This allows you to subclass the AVL tree in the unit test and access the internals.
2. **include the cpp in the test while #def private to public** - I have a colleague that is very smart and does this so it's somehow not as completely awful as it sounds (I guess). You will have to tweak the cmake to use this method since you can't also compile in the cpp.

Reply:
> 2\. **include the cpp in the test while #def private to public**

<sub>But does your colleague know that this pattern has a name: [**public Morozov**](https://en.wikipedia.org/wiki/Pavlik_Morozov)?</sub>

Reply:
It also acts as a 'for each' method, traversing all of the tree entries (future use).

In addition to what I said earlier, there is not a more elegant way of quickly checking the structure; The simple array of references should be swapt out with much more code that would look like `EXPECT_EQ(g, root_->left->right->left->right->left->right...)` (you get the point - it's not that readable).

If you insist, I can do it the `friend` way, but I suggest that we keep this as is, while also acting as an iterator.

Reply:
okay